File: src/modules/Pieces/Kinds/King/King.cpp
#pragma once

#include "King.h"

#include "../../../Common/InstanceOf/InstanceOf.h"

King::King(PieceParams* params):
    CastlePiece(params)
{}

vector<RawPosition> King::getPossibleMoves() {
    return this->getPossibleMoves(false);
}

void King::updatePosition(RawPosition* rawPosition) {
    RawPosition* previousPosition = this->pieceImpl->getRawPosition();
    GameController* gameController = this->pieceImpl->getGameController();
    unique_ptr<GameState>& gameState = gameController->gameControllerImpl->getGameState();

    bool goingToRight = rawPosition->getColumn() > previousPosition->getColumn();

    // if the King moves two squares, this is a castle
    if (abs(previousPosition->getColumn() - rawPosition->getColumn()) == 2) {
            GetValidPieceResponse response = gameState->gameStateImpl->getValidPiece(
                goingToRight ? 7 : 0,
                this->getCorrespondingRow()
            );
        
        if (!response.isOutOfScope && InstanceOf::check<Rook>(response.piece)) {
            RawPosition newRookPosition = RawPosition(
                rawPosition->getColumn() + (goingToRight ? -1 : 1),
                this->getCorrespondingRow()
            );

            static_cast<Rook*>(response.piece)->updatePosition(&newRookPosition);
        }
    }

    // The King calls the super method because it
    // always lands on the correct position when castling!
    Piece::updatePosition(rawPosition);
}

vector<RawPosition> King::getPossibleMoves(bool disableCastle = false) {
    GameController* gameController = this->pieceImpl->getGameController();
    unique_ptr<GameState>& gameState = gameController->gameControllerImpl->getGameState();

    RawPosition* rawPosition = this->pieceImpl->getRawPosition();

    int column = rawPosition->getColumn();
    int row    = rawPosition->getRow();

    // Regular moves
    vector<RawPosition> possibleMoves = vector<RawPosition>();

    possibleMoves.push_back(RawPosition(column, row - 1)); // Top
    possibleMoves.push_back(RawPosition(column, row + 1)); // Bottom
    possibleMoves.push_back(RawPosition(column + 1, row)); // Right
    possibleMoves.push_back(RawPosition(column - 1, row)); // Left

    possibleMoves.push_back(RawPosition(column - 1, row - 1)); // Top left
    possibleMoves.push_back(RawPosition(column + 1, row + 1)); // Top right
    possibleMoves.push_back(RawPosition(column - 1, row + 1)); // Bottom left
    possibleMoves.push_back(RawPosition(column + 1, row - 1)); // Bottom right

    if (!disableCastle) {
        // Castle to the left if possible
        int leftRookOriginalColumn = 0;

        GetValidPieceResponse leftRook = gameState->gameStateImpl->getValidPiece(
            leftRookOriginalColumn,
            this->getCorrespondingRow()
        );

        bool thisIsTreatened = this->isTreatened();

        Rook* leftRookCasted = static_cast<Rook*>(leftRook.piece);

        bool canCastleToLeft =
            !this->getHasMoved()
            && !leftRook.isOutOfScope
            && InstanceOf::check<Rook>(leftRook.piece)
            && !leftRookCasted->getHasMoved()
            && !thisIsTreatened
            && this->checkIfEmptyAndSafe(3, 1) // Check left side
            && !leftRookCasted->isTreatened();

        if (canCastleToLeft) {
            possibleMoves.push_back(RawPosition(
                2,
                this->getCorrespondingRow()
            ));
        }

        // Castle to the right if possible
        bool rightRookOriginalColumn = 7;

        GetValidPieceResponse rightRook = gameState->gameStateImpl->getValidPiece(
            rightRookOriginalColumn,
            this->getCorrespondingRow()
        );
        
        Rook* rightRookCasted = static_cast<Rook*>(rightRook.piece);

        bool canCastleToRight =
            !this->getHasMoved()
            && InstanceOf::check<Rook>(rightRook.piece)
            && !rightRookCasted->getHasMoved()
            && !thisIsTreatened
            && this->checkIfEmptyAndSafe(5, 6) // Check right side
            && !rightRookCasted->isTreatened();
        
        if (canCastleToRight) {
            possibleMoves.push_back(RawPosition(
                6,
                this->getCorrespondingRow()
            ));
        }
    }

    return this->pieceImpl->filterValidMoves(possibleMoves);
}

File: packages/interactive_2d/chess/chess_react/src/Sketches/ChessSketch/elements/Pieces/Kinds/King.ts
/* eslint-disable @typescript-eslint/no-explicit-any */
import { Piece } from "../Piece";
import { CastlePiece } from "../Utils/CastlePiece";
import { Position } from "../types";
import { Rook } from "./Rook";

export class King extends CastlePiece {
    // TODO: Override the method filterValidMoves to
    // also prevent it to moving to an unsafe position

    /**
     * @override
     */
    public updatePosition(rawPosition: Position): void {
        const previousPosition = this.position;
        const newPosition = rawPosition;

        const goingToRight: boolean = newPosition.column > previousPosition.column;

        // if the King moves two squares, this is a castle
        if (Math.abs(previousPosition.column - newPosition.column) === 2) {
            const rook = this.controller.gameState.board
                ?.[this.correspondingRow]
                ?.[goingToRight ? 7 : 0];
            
            if (rook instanceof Rook) {
                const newRookPosition: Position = {
                    row    : this.correspondingRow,
                    column : newPosition.column + (goingToRight ? -1 : 1),
                };

                rook.updatePosition(newRookPosition);
            }

        }

        // The King calls the super method because it
        // always lands on the correcrt position when castling!
        super.updatePosition(rawPosition);
    }

    /**
     * @override
     */
    public getPossibleMoves({ disableCastle } = { disableCastle : false }): Position[] {
        const { column, row } = this.position;

        // Regular moves
        const possibleMoves = [
            { column, row : row - 1 }, // Top
            { column, row : row + 1 }, // Bottom
            { column : column + 1, row }, // Right
            { column : column - 1, row }, // Left

            { column : column - 1, row : row - 1 }, // Top left
            { column : column + 1, row : row + 1 }, // Top right
            { column : column - 1, row : row + 1 }, // Bottom left
            { column : column + 1, row : row - 1 }, // Bottom right
        ] as Position[];

        if (!disableCastle) {
            const { board } = this.controller.gameState;
    
            // Castle to the left if possible
            const leftRookOriginalColumn = 0 as const;

            const leftRook: Rook | Piece | null = board
                ?.[this.correspondingRow]
                ?.[leftRookOriginalColumn];
    
            const thisIsTreatened: boolean = this.isTreatened();

            const canCastleToLeft: boolean =
                !this.hasMoved
                && leftRook instanceof Rook
                && !leftRook.hasMoved
                && !thisIsTreatened
                && this.checkIfEmptyAndSafe(3, 1) // Check left side
                && !leftRook.isTreatened();
    
            if (canCastleToLeft) {
                possibleMoves.push({
                    row    : this.correspondingRow,
                    column : 2,
                });
            }
    
            // Castle to the right if possible
            const rightRookOriginalColumn = 7 as const;

            const rightRook: Rook | Piece | null = board
                ?.[this.correspondingRow]
                ?.[rightRookOriginalColumn];
            
            const canCastleToRight: boolean =
                !this.hasMoved
                && rightRook instanceof Rook
                && !rightRook.hasMoved
                && !thisIsTreatened
                && this.checkIfEmptyAndSafe(5, 6) // Check right side
                && !rightRook.isTreatened();
            
            if (canCastleToRight) {
                possibleMoves.push({
                    row    : this.correspondingRow,
                    column : 6,
                });
            }
        }

        return this.filterValidMoves(possibleMoves);
    }
}
